### PR TITLE
Tabloid web api

### PR DESCRIPTION
**Pull request by askingalot**

### CODE REVIEW

File: book-3-web-api/chapters/ASP_NET_WEB_API_INTRO.md [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+
+Open the console and then click the `Run It!` button. What do you see?
+
+That's right...a "CORS" error.
+
+```txt
+Access to fetch at 'https://localhost:5001/api/beanvariety/' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
+```
+
+[CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors) is a browser security feature that prevents JavaScript from talking to APIs without the web server's concent. CORS is extremely important for production applications, but in development we can afford to be a bit more lax. Update the `Configure` method in the `Startup` class to call `app.UseCors()` to configure CORS behavior.
```
Comment (AdamSheaffer) [inline book-3-web-api/chapters/ASP_NET_WEB_API_INTRO.md:445]: consent (sp)

File: book-3-web-api/chapters/API_OVERVIEW.md [hunk 2]
```diff
@@ -1,18 +1,28 @@
-# ASP.NET Web API
+# Web APIs
 
-In the client side course, you learned how to create a JSON based API using `json-server`. You also may have consumed a third-party API for your client side capstone project. Organizations provide APIs so that software developers can build interesting products based on their data.
+Before we talk about building Web APIs in ASP<span>.NET</span> Core, let's take a moment to talk about APIs more generally.
 
-With ASP.NET, you can use the Web API project type to actually build business logic into your API. `json-server` was severely limited in that it could only create, update, and read JavaScript objects in arrays in your JSON file. In the upcoming sprint, you can perform multiple actions in response to a request from a client.
+An API is an "Application Programming Interface". An API can be compared to a UI (User Interface). A UI consists of components such as textboxes, buttons, menus, paragraphs of information, etc... that a human uses to interact with software. For example, in a Dog Walking MVC application a user may enter dog information into a form and click a submit button in order to sava a new dog to the database.
```
Comment (AdamSheaffer) [inline book-3-web-api/chapters/API_OVERVIEW.md:5]: save (sp)

File: book-3-web-api/chapters/API_OVERVIEW.md [hunk 3]
```diff
@@ -1,18 +1,28 @@
...
+## `json-server`
+
+In the client side course, you learned how to create a JSON based Web API using `json-server`.
+
+Let's consider `json-server` for a moment. It's a pretty impressive application, isn't it? It reads a JSON file and dynamically generates Web API _endpoints_ for the resources in that file. All we have to do to use it is create a JSON file and run `json-server` from the command line. The code in the `json-server` application handles responding to `GET`, `POST`, `PUT`, `PATCH` and `DELETE` requests and performing the appropriate CRUD operations in the JSON file.
+
+As impressive as `json-server` is, however, it has limitations. Firstly, it does not allow us to store our data in a real database. Storing data in a JSON file is great if you're app is only going to be used by one person at a time and doesn't have a lot of data, but once you want to put a web application in production to be used by several people and with real-world data, you need a real database like SQL Server.
```
Comment (AdamSheaffer) [inline book-3-web-api/chapters/API_OVERVIEW.md:17]: your app (sp)

File: book-3-web-api/chapters/API_OVERVIEW.md [hunk 4]
```diff
@@ -1,18 +1,28 @@
...
+
+Secondly, `json-server` will only perform simple CRUD operations to our data. We cannot add _logic_ to the `json-server` API. For example, imagine we'd like to verify that a Dog is not already in the database before we save it. `json-server` will not help us. We will be forced to download the entire list of Dogs and search through it in our JavaScript code. This is, at best, inefficient and, at worst, a security concern since we are potentially downloading sensitive data to the user's browser.
+
+## ASP<span>.NET</span> Core Web API
+
+So how do we overcome the issues of `json-server`? We write our own Web API! Instead of the dynamically generated, but limited `json-server`, we create a custom application that is build specifically for our data and our business rules.
```
Comment (AdamSheaffer) [inline book-3-web-api/chapters/API_OVERVIEW.md:23]: built (sp)

File: book-3-web-api/chapters/ASP_NET_WEB_API_INTRO.md [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+                }
+            }
+        }
+    }
+}
+```
+
+## Controllers
+
+Controllers in Web API are similar to controllers in MVC with a few small differences. The perform the same function in MVC. As in MVC a Web API controller contains methods to respond to HTTP requests.
```
Comment (AdamSheaffer) [inline book-3-web-api/chapters/ASP_NET_WEB_API_INTRO.md:260]: They perform (sp)

File: book-3-web-api/chapters/CONTROLLER_PARAMETERS.md [hunk 6]
```diff
@@ -1,67 +1,91 @@
...
+`_expand` and `_embed` are examples of _query string parameters_.
+
+A _query string_ is the part of a URL after the `?`. It consists of a list of `key=value` pairs separated by `&`.
 
-```sh
-http://localhost:5000/students?cohort=13&orderBy=lastname&limit=5
+Consider this URL:
+
+```txt
+https://www.reddit.com/search/?q=csharp&sort=top&t=month
```
Comment (AdamSheaffer) [inline book-3-web-api/chapters/CONTROLLER_PARAMETERS.md:70]: great example

File: book-3-web-api/chapters/CONTROLLER_PARAMETERS.md [hunk 7]
```diff
@@ -1,67 +1,91 @@
...
+* `q` is the search term
+* `sort` defines the order in which the search results should be returned. In this example we want "top" results to be first...whatever that means.
+* `t` specifies the time limit to search. In this example we only want results from the past month
 
-```cs
-public async Task<IActionResult> Get(int? cohort, string orderBy, int limit)
-```
+The Reddit example demonstrates a common use case for query string parameters. They are often used for searching, sorting and configuring how data is returned from a web API.
+
+> **NOTE:** It is important to note that query string parameters can be whatever you, the developer, want them to be. There is nothing special about the query string parameters in the example. They are just what the Reddit developers chose.
```
Comment (AdamSheaffer) [inline book-3-web-api/chapters/CONTROLLER_PARAMETERS.md:81]: `> **NOTE:** It is important to note that query string parameters can be whatever you, the developer, want them to be. There is nothing special about the query string parameters in the example. They are just what the Reddit developers chose. In the same way, there is nothing special about the _embed or _expand parameters you used with json-server. These are parameter names that the developer of json-server picked out, and are not necessarily relevant keywords outside of json-server`

File: book-3-web-api/chapters/EF_INTRO.md [hunk 8]
```diff
@@ -1,272 +1,480 @@
...
+```
 
-* **Microsoft.EntityFrameworkCore**
-    * The foundational EF library
-* **Microsoft.EntityFrameworkCore.SqlServer**
-    * Specific bits for connecting EF with SQL Server
-* **Microsoft.EntityFrameworkCore.Tool**
-    * Used for working with database "migrations"
- 
+Now go to https://localhost:5001/post/getbyuser/1 to get Posts from the user with id 1.
```
Comment (AdamSheaffer) [inline book-3-web-api/chapters/EF_INTRO.md:411]: I think this ends up being `api/post/getbyuser/1`

File: book-3-web-api/chapters/REACT_WITH_API.md [hunk 9]
```diff
@@ -0,0 +1,210 @@
+# Adding a React Client
+
+The cool thing about exposing a Web API is that any client capable of making HTTP requests can communicate with it. You can create web apps in React, Angular, or Vue and they could all talk to the same API. It even extends past browser based clients. You could make a mobile or desktop app and they could talk to the API. Maybe even IOT devices like microwaves or refrigerators want to make requests to your API.
```
Comment (askingalot) [inline book-3-web-api/chapters/REACT_WITH_API.md:3]: People my not know the acronym. I would spell it out "Internet of Things (IOT)"

File: book-3-web-api/chapters/REACT_WITH_API.md [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+## Create React App
+ 
+Programming refrigerators may come in a later chapter. For now, lets build a react client that will make requests to our API. Open your Gifter project and make a directory called `wwwroot`. This is where we'll put our React code. In your terminal, `cd` into the wwwroot directory and run `npx create-react-app client`
```
Comment (askingalot) [inline book-3-web-api/chapters/REACT_WITH_API.md:7]: I think the `client` folder should be outside of `wwwroot`. Maybe later we can teach them to build to `wwwroot`.

File: book-3-web-api/chapters/REACT_WITH_API.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    </PostContext.Provider>
+  );
+};
+```
+
+This providers the state value of the posts array, as well as methods to fetch all posts and add a new post. Note that the urls we are making requests to are relative urls--they don't have anything like `https://localhost:5001/api/posts`. This is a benefit of adding the `proxy` attribute in our package.json file.
+
+## Adding a Posts List Component
+
+Inside the components director, create a file called `PostList.js` and add the following code
```
Comment (askingalot) [inline book-3-web-api/chapters/REACT_WITH_API.md:63]: "components director**y**"

File: book-3-web-api/chapters/REACT_WITH_API.md [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+        </div>
+      </div>
+    </div>
+  );
+};
+
+export default PostList;
+```
+
+## Exercise
```
Comment (askingalot) [inline book-3-web-api/chapters/REACT_WITH_API.md:193]: I think we should add an exercise to have them list all the comments beneath each post. We can refactor after we do routing to have them only show comments on the Post "details" page, but I think another display-only component will help them get back into the swing of react.

I'm assuming we'll have a comments array on each post.

**Pull request merged**